pkg/collector/corechecks/snmp/checkconfig/config.go

```diff
@@ -0,0 +1,482 @@
+		if err != nil {
+			return nil, fmt.Errorf("failed to load default profiles: %s", err)
+		}
+		profiles = defaultProfiles
+	}
+
+	for _, profileDef := range profiles {
+		normalizeMetrics(profileDef.Metrics)
+	}
+
+	c.Profiles = profiles
+	profile := instance.Profile
+
+	errors := validateEnrichMetrics(c.Metrics)
+	errors = append(errors, ValidateEnrichMetricTags(c.MetricTags)...)
+	if len(errors) > 0 {
+		return nil, fmt.Errorf("validation errors: %s", strings.Join(errors, "\n"))
+	}
+
+	if profile != "" || len(c.Metrics) > 0 {
+		c.AutodetectProfile = false
+	} else {
+		c.AutodetectProfile = true
+	}
+
+	if profile != "" {
+		err = c.RefreshWithProfile(profile)
+		if err != nil {
+			return nil, fmt.Errorf("failed to refresh with profile `%s`: %s", profile, err)
+		}
+	}
+
+	c.DeviceID, c.DeviceIDTags = buildDeviceID(c.getDeviceIDTags())
+
+	subnet, err := getSubnetFromTags(c.InstanceTags)
+	if err != nil {
+		log.Debugf("subnet not found: %s", err)
+	}
+	c.Subnet = subnet
+
+	c.addUptimeMetric()
+	return c, nil
+}
+
+// Copy makes a copy of CheckConfig
+func (c *CheckConfig) Copy() *CheckConfig {
+	newConfig := CheckConfig{}
+	newConfig.IPAddress = c.IPAddress
+	newConfig.Port = c.Port
+	newConfig.CommunityString = c.CommunityString
+	newConfig.SnmpVersion = c.SnmpVersion
+	newConfig.Timeout = c.Timeout
+	newConfig.Retries = c.Retries
+	newConfig.User = c.User
+	newConfig.AuthProtocol = c.AuthProtocol
+	newConfig.AuthKey = c.AuthKey
+	newConfig.PrivProtocol = c.PrivProtocol
+	newConfig.PrivKey = c.PrivKey
+	newConfig.ContextName = c.ContextName
+	newConfig.ContextName = c.ContextName
+	newConfig.OidConfig = c.OidConfig
+	newConfig.Metrics = make([]MetricsConfig, 0, len(c.Metrics))
+	for _, metric := range c.Metrics {
+		newConfig.Metrics = append(newConfig.Metrics, metric)
+	}
+	newConfig.MetricTags = make([]MetricTagConfig, 0, len(c.MetricTags))
+	for _, metricTag := range c.MetricTags {
+		newConfig.MetricTags = append(newConfig.MetricTags, metricTag)
+	}
+	newConfig.OidBatchSize = c.OidBatchSize
+	newConfig.BulkMaxRepetitions = c.BulkMaxRepetitions
+	newConfig.Profiles = c.Profiles
+	newConfig.ProfileTags = common.CopyStrings(c.ProfileTags)
+	newConfig.Profile = c.Profile
+	newConfig.ProfileDef = c.ProfileDef
+	newConfig.ExtraTags = common.CopyStrings(c.ExtraTags)
+	newConfig.InstanceTags = common.CopyStrings(c.InstanceTags)
+	newConfig.CollectDeviceMetadata = c.CollectDeviceMetadata
+	newConfig.DeviceID = c.DeviceID
+
+	newConfig.DeviceIDTags = common.CopyStrings(c.DeviceIDTags)
+	newConfig.Subnet = c.Subnet
+	newConfig.AutodetectProfile = c.AutodetectProfile
+	newConfig.MinCollectionInterval = c.MinCollectionInterval
+
+	return &newConfig
+}
+
+// CopyWithNewIP makes a copy of CheckConfig with new IP
+func (c *CheckConfig) CopyWithNewIP(ipAddress string) *CheckConfig {
+	newConfig := c.Copy()
+	newConfig.IPAddress = ipAddress
+	newConfig.DeviceID, newConfig.DeviceIDTags = buildDeviceID(newConfig.getDeviceIDTags())
```

Comment:
> could it be a method within the config to UpdateDeviceID ?
Good idea. Changed the code here https://github.com/DataDog/datadog-agent/pull/8971/commits/a7f71a33d8bf32885c06118fd8d0f39d20a911d1

> What if DeviceID and DeviceIDTags were methods ?
Not sure about this, since it will lead to DeviceID to be hashed and new tags being created many times (see buildDeviceID function).